Improve spacing of gallery list card and dialog text

mkx173 opened the pull request:
优化画廊及下载页面卡片内容的对齐
上：修改前
下：修改后

![Screenshot_20240416_203116_EhViewer Debug](https://github.com/FooIbar/EhViewer/assets/13722982/4133ea30-1e7a-4129-9bef-e1edab427c58)
![Screenshot_20240416_203136_EhViewer Debug](https://github.com/FooIbar/EhViewer/assets/13722982/54eee2eb-2d5d-43c3-a656-42ce6a62e01d)

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/DownloadCard.kt:
```diff
@@ -91,7 +95,7 @@ fun DownloadCard(
                 else -> null // Chill, will be removed soon
             }
-            ConstraintLayout(modifier = Modifier.padding(8.dp, 4.dp).fillMaxSize()) {
+            ConstraintLayout(modifier = Modifier.padding(8.dp, 0.dp).fillMaxSize()) {
```
`Modifier.padding(horizontal: Dp = 8.dp)`

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/GalleryInfo.kt:
```diff
@@ -110,7 +113,7 @@ fun GalleryInfoListItem(
             )
         }
-        ConstraintLayout(modifier = Modifier.padding(8.dp, 4.dp).fillMaxSize(), constraintSet = constraintSet) {
+        ConstraintLayout(modifier = Modifier.padding(8.dp, 0.dp).fillMaxSize(), constraintSet = constraintSet) {
```
Ditto.

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/GalleryInfo.kt:
```diff
@@ -153,13 +157,15 @@ fun GalleryInfoListItem(
                     }
-                    Text(text = info.simpleLanguage.orEmpty())
+                    if (info.simpleLanguage != null) {
+                        Text(text = info.simpleLanguage.orEmpty())
```
`info.simpleLanguage?.let {}`

revonateB0T replied:
`Text(text = it)`

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/DownloadCard.kt:
```diff
@@ -173,12 +178,13 @@ fun DownloadCard(
                         end.linkTo(parent.end)
                         bottom.linkTo(parent.bottom)
-                    },
+                    }.then(Modifier.size(actionButtonSize).offset(x = 8.dp))
```
What's the explicit `then` for?

mkx173 replied:
This was needed in a previous version to bypass the 48dp minimum size requirement for IconButton. No longer needed now.

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/DownloadCard.kt:
```diff
@@ -119,12 +123,13 @@ fun DownloadCard(
                                 overflow = TextOverflow.Ellipsis,
                             )
                         }
+                        val ratingBasePadding = (LocalTextStyle.current.lineHeight.value.dp - dimensionResource(id = R.dimen.rating_size)) / 2
```
`TextStyle.lineHeight`'s value is `sp`, we shouldn't use it as `dp` directly.
Find a better approach.

mkx173 replied:
```kotlin
with(LocalDensity.current) {
    LocalTextStyle.current.fontSize.toDp()
}
```
Will use something like this instead.

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/DownloadCard.kt:
```diff
@@ -119,12 +123,13 @@ fun DownloadCard(
                                 start.linkTo(parent.start)
                                 bottom.linkTo(categoryRef.top)
-                            },
+                            }.padding(top = ratingBasePadding - 2.dp, bottom = ratingBasePadding + 2.dp),
```
Can we configure the padding with `constrainAs` scope?

mkx173 commented:
In the new commits I modified the RatingWidget to make rating icons scale with font size in gallery and download card. The action button was also modified to scale accordingly.

For the size of the action button, I added MeasureUnconstrainedViewHeight to measure the height of a Text composable and set the height of the action button according to it. This was to make sure the vertical spacing of the text and icons remain consistent across gallery and download page. Not sure if there are better ways though.

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/DownloadCard.kt:
```diff
@@ -173,40 +184,76 @@ fun DownloadCard(
+
+// https://stackoverflow.com/questions/70500071/is-it-possible-to-measure-string-width-to-properly-size-a-text-composable
+@Composable
+internal fun MeasureUnconstrainedViewHeight(
```
I would like to avoid this
Can't we just calculate the height(dp) by `fontScale` and `labelLarge` fontSize?

revonateB0T replied:
@mkx173 
https://github.com/revonateB0T/EhViewer/commit/b15337bd531778f692691b09a01c76b3b0f325f2
Does it work as intended for you?

revonateB0T replied:
Also see https://github.com/revonateB0T/EhViewer/commit/4fffa045166a78a952f92484026cb5a3e6aa9cb0

mkx173 replied:
Yes this seems to work well here. I cherry picked your commits to this pull request.

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/DownloadCard.kt:
```diff
@@ -173,40 +184,76 @@ fun DownloadCard(
+                        val running =
+                            downloadState == DownloadInfo.STATE_WAIT || downloadState == DownloadInfo.STATE_DOWNLOAD
+                        val icon = remember {
+                            movableContentOf<Boolean> {
```
@FooIbar 
Why the `Icon` needs to be inside `movableContentOf`?

FooIbar replied:
To prevent recomposition, it's not required tho, unlike the thumb.

revonateB0T commented on app/src/main/kotlin/com/hippo/ehviewer/ui/main/GalleryInfo.kt:
```diff
@@ -130,13 +133,16 @@ fun GalleryInfoListItem(
                 }
                 GalleryListCardRating(
                     rating = info.rating,
+                    ratingSize = with(LocalDensity.current) {
```
Should we make this intrinsic for `GalleryListCardRating`, others LGTM.

FooIbar commented:
Rating stars look too small to me.
| | |
| --- | --- |
| Before | ![Screenshot_20240418-115621_Paranoid_Launcher 1](https://github.com/FooIbar/EhViewer/assets/118464521/625094ba-ace1-4e9b-b275-0be8483a64e6) |
| After | ![Screenshot_20240418-115601_Paranoid_Launcher 1](https://github.com/FooIbar/EhViewer/assets/118464521/7c1b4264-bb0c-499e-b817-2977aa87b779) |

mkx173 commented:
> Rating stars look too small to me.
> 
> Before	![Screenshot_20240418-115621_Paranoid_Launcher 1](https://private-user-images.githubusercontent.com/118464521/323461244-625094ba-ace1-4e9b-b275-0be8483a64e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0MTQ1NDYsIm5iZiI6MTcxMzQxNDI0NiwicGF0aCI6Ii8xMTg0NjQ1MjEvMzIzNDYxMjQ0LTYyNTA5NGJhLWFjZTEtNGU5Yi1iMjc1LTBiZTg0ODNhNjRlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxOFQwNDI0MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NmE4MmVmODEyOTJjYzQ2MWZlYWU3NWZmZjViNTAyMTZhMThmNzg5YjdiZjMxMDdkZGFiM2JjMTA4YmQ1M2VjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.jHiPu16_yjBresvsu3dAmMN5GvaG_uiB6P1fM0mBke4)
> After	![Screenshot_20240418-115601_Paranoid_Launcher 1](https://private-user-images.githubusercontent.com/118464521/323461342-7c1b4264-bb0c-499e-b817-2977aa87b779.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0MTQ1NDYsIm5iZiI6MTcxMzQxNDI0NiwicGF0aCI6Ii8xMTg0NjQ1MjEvMzIzNDYxMzQyLTdjMWI0MjY0LWJiMGMtNDk5ZS1iODE3LTI5NzdhYTg3Yjc3OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxOFQwNDI0MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNDBiMGI3ODYzMzFiMzc3N2QzMDg5OThlYzUzMDBiMmI5OTMwNjQ3N2U1ZGI0M2U1ZmJkYTBmODQzNDMxYmFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.XKsZURRReNWXCCWqYB4wLzSYd_fpNsWf-lJem8wHO2A)

Scaled rating icons by 1.1x. Also made download and favorite icons scale with font size.

FooIbar commented:
Kinda not like the idea of multiplying with `measuredHeight`, is 92c54cb what you are trying to achieve?
| | |
| --- | --- |
| Before | ![Screenshot_20240418-164901_Paranoid_Launcher](https://github.com/FooIbar/EhViewer/assets/118464521/3c6b4f4a-cff2-409e-8319-7c0f6deb724c) |
| After | ![Screenshot_20240418-164804_Paranoid_Launcher](https://github.com/FooIbar/EhViewer/assets/118464521/d5584dab-0a53-41ce-9ca3-4db5cacdc973) |

mkx173 commented:
The motivation for the proposed changes is to improve the alignment of text and icon layout and make them consistent across different font and/or UI scaling levels. Accessibility-wise, minimum height for the IconButton will always be reserved even when font size is small, so I would say this is not an issue.

<img width="136" alt="屏幕截图 2024-04-18 030449" src="https://github.com/FooIbar/EhViewer/assets/13722982/0691c9fc-9052-422a-8499-8fc8be9f01e5">

![Screenshot_20240418_031040](https://github.com/FooIbar/EhViewer/assets/13722982/3e41c32a-785a-466e-a341-066447ddf96f)
![Screenshot_20240418_030923](https://github.com/FooIbar/EhViewer/assets/13722982/ff48a174-6a7b-4b06-a55d-7a108d1418e9)

FooIbar commented:
![Screenshot_20240418-184910_Paranoid_Launcher](https://github.com/FooIbar/EhViewer/assets/118464521/26051c2e-0a00-471e-9dfd-619dc90239c4)
It's not aligned in the current implementation though.
The state text overlaps the action button, and the rating stars are closer to the uploader text than the category text.
![Screenshot_20240418-190139_Paranoid_Launcher](https://github.com/FooIbar/EhViewer/assets/118464521/b5cf9ede-1181-44b3-b672-f19581cf94d7)
And it doesn't handle the case where page numbers are not shown.
You can take what I posted earlier as reference.

mkx173 commented:
| |  |
|--------|--------|
| With 4dp margin | ![Screenshot_20240418_043922](https://github.com/FooIbar/EhViewer/assets/13722982/8d58d615-5a95-4016-8503-a29fae0b6ef5) |
| Without margin  | ![Screenshot_20240418_043936](https://github.com/FooIbar/EhViewer/assets/13722982/37c76d38-f71e-4976-a913-2427c0981b66) |

The 4dp margin is intended to make rating icons visually aligned between the two lines. There will still be a small difference due to different "baseline" levels of English and CJK characters, but I think 4dp is a good balance point.

The second issue is fixed in 0e9e440. Thanks for pointing out.

mkx173 commented:
<img width="901" alt="image" src="https://github.com/FooIbar/EhViewer/assets/13722982/0641e81c-a47a-4139-8ee0-517295cefa08">

Also from this image you can see why the state text may overlap with the action button. In gallery page the uploader text is aligned with favorite text, which in turns is stacked on the date text. However, in download page we do not have those two, so we have to calculate the height of the action button to keep uploader text consistent with gallery page. When the text size is small, the touch zone of the action button will overlap with the state text to ensure minimum size.

FooIbar commented:
Sorry, I don't like icons to be scaled and we didn't do that for other icons.
The alignment tweak will be implemented as part of #1036.
